File: sdks/arduino/MaestraClient/src/MaestraClient.cpp
/**
 * Maestra Client Implementation
 */

#include "MaestraClient.h"

// Global client pointer for callback
static MaestraClient* _globalClient = nullptr;

// MQTT callback wrapper
static void mqttCallback(char* topic, byte* payload, unsigned int length) {
    if (_globalClient) {
        _globalClient->_handleMessage(topic, payload, length);
    }
}

// ============================================================================
// MaestraEntityState
// ============================================================================

MaestraEntityState::MaestraEntityState() {
    _state = _doc.to<JsonObject>();
}

bool MaestraEntityState::has(const char* key) {
    return _state.containsKey(key);
}

JsonObject MaestraEntityState::data() {
    return _state;
}

void MaestraEntityState::_update(JsonObject newState) {
    for (JsonPair kv : newState) {
        _state[kv.key()] = kv.value();
    }
}

// ============================================================================
// MaestraEntity
// ============================================================================

MaestraEntity::MaestraEntity(MaestraClient* client, const char* slug)
    : _client(client), _callback(nullptr) {
    strncpy(_slug, slug, sizeof(_slug) - 1);
    _slug[sizeof(_slug) - 1] = '\0';
}

void MaestraEntity::updateState(JsonObject updates) {
    _client->updateEntityState(_slug, updates);
}

void MaestraEntity::updateState(const char* key, JsonVariant value) {
    StaticJsonDocument<256> doc;
    JsonObject obj = doc.to<JsonObject>();
    obj[key] = value;
    updateState(obj);
}

void MaestraEntity::setState(JsonObject newState) {
    _client->setEntityState(_slug, newState);
}

void MaestraEntity::onStateChange(StateChangeCallback callback) {
    _callback = callback;
}

void MaestraEntity::_handleMessage(JsonObject payload) {
    if (!payload.containsKey("current_state")) return;

    JsonObject currentState = payload["current_state"];
    _state._update(currentState);

    if (_callback) {
        JsonArray changedKeys = payload["changed_keys"];
        _callback(_slug, currentState, changedKeys);
    }
}

// ============================================================================
// MaestraClient
// ============================================================================

MaestraClient::MaestraClient(Client& networkClient)
    : _mqtt(networkClient), _port(1883), _hasCredentials(false),
      _entityCount(0), _streamCallback(nullptr), _streamCount(0) {
    strcpy(_broker, "localhost");
    strcpy(_clientId, "maestra-arduino");
    _globalClient = this;

    for (int i = 0; i < MAX_ENTITIES; i++) {
        _entities[i] = nullptr;
    }
    memset(_streams, 0, sizeof(_streams));
}

void MaestraClient::setBroker(const char* host, uint16_t port) {
    strncpy(_broker, host, sizeof(_broker) - 1);
    _port = port;
}

void MaestraClient::setClientId(const char* clientId) {
    strncpy(_clientId, clientId, sizeof(_clientId) - 1);
}

void MaestraClient::setCredentials(const char* username, const char* password) {
    strncpy(_username, username, sizeof(_username) - 1);
    strncpy(_password, password, sizeof(_password) - 1);
    _hasCredentials = true;
}

bool MaestraClient::connect() {
    _mqtt.setServer(_broker, _port);
    _mqtt.setCallback(mqttCallback);
    _mqtt.setBufferSize(MAESTRA_JSON_BUFFER_SIZE);

    bool connected;
    if (_hasCredentials) {
        connected = _mqtt.connect(_clientId, _username, _password);
    } else {
        connected = _mqtt.connect(_clientId);
    }

    if (connected) {
        Serial.println("✅ Connected to Maestra MQTT broker");
    } else {
        Serial.print("❌ MQTT connection failed, rc=");
        Serial.println(_mqtt.state());
    }

    return connected;
}

void MaestraClient::disconnect() {
    _mqtt.disconnect();
}

bool MaestraClient::isConnected() {
    return _mqtt.connected();
}

void MaestraClient::loop() {
    _mqtt.loop();
}

MaestraEntity* MaestraClient::getEntity(const char* slug) {
    // Check if already exists
    for (int i = 0; i < _entityCount; i++) {
        if (_entities[i] && strcmp(_entities[i]->slug(), slug) == 0) {
            return _entities[i];
        }
    }

    // Create new entity
    if (_entityCount < MAX_ENTITIES) {
        MaestraEntity* entity = new MaestraEntity(this, slug);
        _entities[_entityCount++] = entity;
        return entity;
    }

    return nullptr;
}

void MaestraClient::subscribeEntity(const char* slug) {
    char topic[MAESTRA_TOPIC_BUFFER_SIZE];
    snprintf(topic, sizeof(topic), "maestra/entity/state/+/%s", slug);
    _mqtt.subscribe(topic);
    Serial.print("📡 Subscribed to: ");
    Serial.println(topic);
}

void MaestraClient::updateEntityState(const char* slug, JsonObject state, const char* source) {
    _publishState(slug, state, source, false);
}

void MaestraClient::setEntityState(const char* slug, JsonObject state, const char* source) {
    _publishState(slug, state, source, true);
}

void MaestraClient::_publishState(const char* slug, JsonObject state, const char* source, bool replace) {
    char topic[MAESTRA_TOPIC_BUFFER_SIZE];
    snprintf(topic, sizeof(topic), "maestra/entity/state/%s/%s",
             replace ? "set" : "update", slug);

    StaticJsonDocument<MAESTRA_JSON_BUFFER_SIZE> doc;
    doc["state"] = state;
    if (source) {
        doc["source"] = source;
    }

    char buffer[MAESTRA_JSON_BUFFER_SIZE];
    serializeJson(doc, buffer);

    _mqtt.publish(topic, buffer);
}

// ============================================================================
// Stream Methods
// ============================================================================

void MaestraClient::subscribeStreamEvents(StreamAdvertisedCallback callback) {
    _streamCallback = callback;
    _mqtt.subscribe("maestra/stream/advertise");
    Serial.println("📡 Subscribed to stream events");
}

void MaestraClient::subscribeStreamType(const char* streamType, StreamAdvertisedCallback callback) {
    _streamCallback = callback;
    char topic[MAESTRA_TOPIC_BUFFER_SIZE];
    snprintf(topic, sizeof(topic), "maestra/stream/advertise/%s", streamType);
    _mqtt.subscribe(topic);
    Serial.print("📡 Subscribed to stream type: ");
    Serial.println(streamType);
}

void MaestraClient::advertiseStream(const char* name, const char* streamType,
                                     const char* protocol, const char* address,
                                     int port, const char* publisherId) {
    StaticJsonDocument<MAESTRA_JSON_BUFFER_SIZE> doc;
    doc["name"] = name;
    doc["stream_type"] = streamType;
    doc["publisher_id"] = publisherId ? publisherId : _clientId;
    doc["protocol"] = protocol;
    doc["address"] = address;
    doc["port"] = port;

    char buffer[MAESTRA_JSON_BUFFER_SIZE];
    serializeJson(doc, buffer);

    _mqtt.publish("maestra/stream/advertise", buffer);
    Serial.print("📡 Advertised stream: ");
    Serial.println(name);
}

void MaestraClient::withdrawStream(const char* streamId) {
    char topic[MAESTRA_TOPIC_BUFFER_SIZE];
    snprintf(topic, sizeof(topic), "maestra/stream/withdraw/%s", streamId);
    _mqtt.publish(topic, "{}");
}

void MaestraClient::streamHeartbeat(const char* streamId) {
    char topic[MAESTRA_TOPIC_BUFFER_SIZE];
    snprintf(topic, sizeof(topic), "maestra/stream/heartbeat/%s", streamId);
    _mqtt.publish(topic, "{}");
}

void MaestraClient::_handleStreamMessage(const char* topic, JsonObject payload) {
    if (!_streamCallback) return;

    const char* streamId = payload["id"] | "";
    const char* streamName = payload["name"] | "";
    const char* streamType = payload["stream_type"] | "";
    const char* address = payload["address"] | "";
    int port = payload["port"] | 0;

    _streamCallback(streamId, streamName, streamType, address, port);
}

// ============================================================================
// Message Handler
// ============================================================================

void MaestraClient::_handleMessage(char* topic, byte* payload, unsigned int length) {
    // Parse topic into parts
    char* parts[6];
    int partCount = 0;
    char topicCopy[MAESTRA_TOPIC_BUFFER_SIZE];
    strncpy(topicCopy, topic, sizeof(topicCopy));
    topicCopy[sizeof(topicCopy) - 1] = '\0';

    char* token = strtok(topicCopy, "/");
    while (token && partCount < 6) {
        parts[partCount++] = token;
        token = strtok(nullptr, "/");
    }

    if (partCount < 3) return;

    // Parse JSON payload
    StaticJsonDocument<MAESTRA_JSON_BUFFER_SIZE> doc;
    DeserializationError error = deserializeJson(doc, payload, length);
    if (error) return;

    // Route stream messages: maestra/stream/advertise[/type]
    if (partCount >= 3 && strcmp(parts[1], "stream") == 0 &&
        strcmp(parts[2], "advertise") == 0) {
        _handleStreamMessage(topic, doc.as<JsonObject>());
        return;
    }

    // Route entity messages: maestra/entity/state/<type>/<slug>
    if (partCount >= 5 && strcmp(parts[1], "entity") == 0) {
        const char* entitySlug = parts[4];

        for (int i = 0; i < _entityCount; i++) {
            if (_entities[i] && strcmp(_entities[i]->slug(), entitySlug) == 0) {
                _entities[i]->_handleMessage(doc.as<JsonObject>());
                break;
            }
        }
    }
}

File: sdks/processing/MaestraClient/src/maestra/MaestraClient.java
package maestra;

import processing.core.PApplet;
import processing.data.JSONObject;
import processing.data.JSONArray;
import mqtt.MQTTClient;
import java.util.ArrayList;
import java.util.concurrent.ConcurrentLinkedQueue;

/**
 * Maestra Client for Processing
 *
 * MQTT-based state management for creative coding with Processing.
 * Requires the processing-mqtt library by 256dpi.
 *
 * Usage:
 *   MaestraClient maestra;
 *
 *   void setup() {
 *       maestra = new MaestraClient(this);
 *       maestra.setBroker("192.168.1.100");
 *       maestra.setClientId("my-sketch");
 *       maestra.connect();
 *   }
 *
 *   void draw() {
 *       maestra.update();  // Required - processes incoming messages
 *   }
 */
public class MaestraClient {
    private PApplet parent;
    private MQTTClient mqtt;
    private String broker = "localhost";
    private int port = 1883;
    private String clientId = "maestra-processing";
    private String username;
    private String password;
    private boolean hasCredentials = false;
    private boolean connected = false;

    // Entity registry
    private ArrayList<MaestraEntity> entities = new ArrayList<MaestraEntity>();

    // Wildcard entity subscriptions
    private WildcardEntityCallback wildcardAllCallback;
    private ArrayList<String> wildcardTypes = new ArrayList<String>();
    private ArrayList<WildcardEntityCallback> wildcardTypeCallbacks = new ArrayList<WildcardEntityCallback>();

    // Stream subscriptions
    private StreamAdvertisedCallback streamCallback;

    // Thread-safe message queue (MQTT callbacks arrive on Paho thread)
    private ConcurrentLinkedQueue<QueuedMessage> messageQueue = new ConcurrentLinkedQueue<QueuedMessage>();

    /**
     * Create a new MaestraClient.
     * @param parent  The Processing PApplet (pass 'this' from your sketch)
     */
    public MaestraClient(PApplet parent) {
        this.parent = parent;
        // Register dispose callback for cleanup on sketch exit
        parent.registerMethod("dispose", this);
    }

    // ========================================================================
    // Configuration
    // ========================================================================

    /** Set the MQTT broker host and port. */
    public void setBroker(String host, int port) {
        this.broker = host;
        this.port = port;
    }

    /** Set the MQTT broker host (default port 1883). */
    public void setBroker(String host) {
        setBroker(host, 1883);
    }

    /** Set the MQTT client ID. */
    public void setClientId(String clientId) {
        this.clientId = clientId;
    }

    /** Set MQTT credentials for authenticated brokers. */
    public void setCredentials(String username, String password) {
        this.username = username;
        this.password = password;
        this.hasCredentials = true;
    }

    // ========================================================================
    // Connection
    // ========================================================================

    /** Connect to the MQTT broker. */
    public boolean connect() {
        try {
            mqtt = new MQTTClient(parent);
            if (hasCredentials) {
                mqtt.connect("mqtt://" + broker + ":" + port, clientId, username, password);
            } else {
                mqtt.connect("mqtt://" + broker + ":" + port, clientId);
            }
            connected = true;
            PApplet.println("[Maestra] Connected to MQTT broker at " + broker + ":" + port);
            return true;
        } catch (Exception e) {
            PApplet.println("[Maestra] Connection failed: " + e.getMessage());
            connected = false;
            return false;
        }
    }

    /** Disconnect from the MQTT broker. */
    public void disconnect() {
        if (mqtt != null) {
            mqtt.disconnect();
            connected = false;
        }
    }

    /** Check if connected to the MQTT broker. */
    public boolean isConnected() {
        return connected;
    }

    /**
     * Process incoming MQTT messages. MUST be called in draw().
     * Drains the message queue and dispatches to entity callbacks on the main thread.
     */
    public void update() {
        QueuedMessage msg;
        while ((msg = messageQueue.poll()) != null) {
            _processMessage(msg.topic, msg.payload);
        }
    }

    /** Called by Processing on sketch exit. */
    public void dispose() {
        disconnect();
    }

    // ========================================================================
    // Entity Management
    // ========================================================================

    /** Get or create an entity reference by slug. */
    public MaestraEntity getEntity(String slug) {
        for (MaestraEntity entity : entities) {
            if (entity.slug().equals(slug)) {
                return entity;
            }
        }
        MaestraEntity entity = new MaestraEntity(this, slug);
        entities.add(entity);
        return entity;
    }

    /** Subscribe to state changes for a specific entity. */
    public void subscribeEntity(String slug) {
        if (mqtt == null) return;
        String topic = "maestra/entity/state/+/" + slug;
        mqtt.subscribe(topic);
        PApplet.println("[Maestra] Subscribed to: " + topic);
    }

    /** Subscribe to state changes for ALL entities. */
    public void subscribeAllEntities(WildcardEntityCallback callback) {
        this.wildcardAllCallback = callback;
        if (mqtt == null) return;
        mqtt.subscribe("maestra/entity/state/#");
        PApplet.println("[Maestra] Subscribed to all entity state changes");
    }

    /** Subscribe to state changes for a specific entity type. */
    public void subscribeEntityType(String type, WildcardEntityCallback callback) {
        wildcardTypes.add(type);
        wildcardTypeCallbacks.add(callback);
        if (mqtt == null) return;
        String topic = "maestra/entity/state/" + type + "/+";
        mqtt.subscribe(topic);
        PApplet.println("[Maestra] Subscribed to entity type: " + type);
    }

    // ========================================================================
    // State Publishing
    // ========================================================================

    /** Publish a state update (merge) for an entity. */
    public void updateEntityState(String slug, JSONObject state, String source) {
        _publishState(slug, state, source, false);
    }

    /** Publish a complete state replacement for an entity. */
    public void setEntityState(String slug, JSONObject state, String source) {
        _publishState(slug, state, source, true);
    }

    private void _publishState(String slug, JSONObject state, String source, boolean replace) {
        if (mqtt == null) return;

        String action = replace ? "set" : "update";
        String topic = "maestra/entity/state/" + action + "/" + slug;

        JSONObject payload = new JSONObject();
        payload.put("state", state);
        if (source != null) {
            payload.put("source", source);
        } else {
            payload.put("source", clientId);
        }

        mqtt.publish(topic, payload.toString());
    }

    // ========================================================================
    // Stream Support
    // ========================================================================

    /** Subscribe to all stream advertisement events. */
    public void subscribeStreamEvents(StreamAdvertisedCallback callback) {
        this.streamCallback = callback;
        if (mqtt == null) return;
        mqtt.subscribe("maestra/stream/advertise");
        PApplet.println("[Maestra] Subscribed to stream events");
    }

    /** Subscribe to stream advertisements of a specific type. */
    public void subscribeStreamType(String streamType, StreamAdvertisedCallback callback) {
        this.streamCallback = callback;
        if (mqtt == null) return;
        String topic = "maestra/stream/advertise/" + streamType;
        mqtt.subscribe(topic);
        PApplet.println("[Maestra] Subscribed to stream type: " + streamType);
    }

    /** Advertise a stream. */
    public void advertiseStream(String name, String streamType, String protocol,
                                 String address, int port) {
        advertiseStream(name, streamType, protocol, address, port, null);
    }

    /** Advertise a stream with a custom publisher ID. */
    public void advertiseStream(String name, String streamType, String protocol,
                                 String address, int port, String publisherId) {
        if (mqtt == null) return;

        JSONObject payload = new JSONObject();
        payload.put("name", name);
        payload.put("stream_type", streamType);
        payload.put("publisher_id", publisherId != null ? publisherId : clientId);
        payload.put("protocol", protocol);
        payload.put("address", address);
        payload.put("port", port);

        mqtt.publish("maestra/stream/advertise", payload.toString());
        PApplet.println("[Maestra] Advertised stream: " + name);
    }

    /** Withdraw (remove) a stream advertisement. */
    public void withdrawStream(String streamId) {
        if (mqtt == null) return;
        mqtt.publish("maestra/stream/withdraw/" + streamId, "{}");
    }

    /** Send a heartbeat for a stream to keep it alive. */
    public void streamHeartbeat(String streamId) {
        if (mqtt == null) return;
        mqtt.publish("maestra/stream/heartbeat/" + streamId, "{}");
    }

    // ========================================================================
    // MQTT Message Handling
    // ========================================================================

    /**
     * Called by processing-mqtt on the Paho thread.
     * The processing-mqtt library calls messageReceived() on the PApplet,
     * but we use clientReceived() pattern to route through this client.
     * Users should add this to their sketch:
     *
     *   void clientConnected() { }
     *   void messageReceived(String topic, byte[] payload) {
     *       maestra.messageReceived(topic, new String(payload));
     *   }
     *   void connectionLost() { maestra.connectionLost(); }
     */
    public void messageReceived(String topic, String payload) {
        // Queue for processing on the main thread
        messageQueue.add(new QueuedMessage(topic, payload));
    }

    /** Called when the MQTT connection is lost. */
    public void connectionLost() {
        connected = false;
        PApplet.println("[Maestra] Connection lost");
    }

    /** Internal: process a message on the main thread (called from update()). */
    private void _processMessage(String topic, String payloadStr) {
        String[] parts = topic.split("/");
        if (parts.length < 3) return;

        JSONObject payload;
        try {
            payload = JSONObject.parse(payloadStr);
        } catch (Exception e) {
            return;
        }
        if (payload == null) return;

        // Route stream messages: maestra/stream/advertise[/type]
        if (parts.length >= 3 && parts[1].equals("stream") && parts[2].equals("advertise")) {
            _handleStreamMessage(payload);
            return;
        }

        // Route entity messages: maestra/entity/state/<type>/<slug>
        if (parts.length >= 5 && parts[1].equals("entity") && parts[2].equals("state")) {
            String entityType = parts[3];
            String entitySlug = parts[4];

            // Dispatch to specific MaestraEntity if registered
            for (MaestraEntity entity : entities) {
                if (entity.slug().equals(entitySlug)) {
                    entity._handleMessage(payload);
                    break;
                }
            }

            // Extract state and changed_keys for wildcard callbacks
            JSONObject currentState = payload.hasKey("current_state")
                ? payload.getJSONObject("current_state") : new JSONObject();
            JSONArray changedKeys = payload.hasKey("changed_keys")
                ? payload.getJSONArray("changed_keys") : new JSONArray();

            // Fire wildcard all-entities callback
            if (wildcardAllCallback != null) {
                wildcardAllCallback.stateChanged(entityType, entitySlug, currentState, changedKeys);
            }

            // Fire matching per-type wildcard callbacks
            for (int i = 0; i < wildcardTypes.size(); i++) {
                if (wildcardTypes.get(i).equals(entityType)) {
                    wildcardTypeCallbacks.get(i).stateChanged(entityType, entitySlug, currentState, changedKeys);
                }
            }
        }
    }

    private void _handleStreamMessage(JSONObject payload) {
        if (streamCallback == null) return;

        String streamId = payload.getString("id", "");
        String name = payload.getString("name", "");
        String streamType = payload.getString("stream_type", "");
        String address = payload.getString("address", "");
        int port = payload.getInt("port", 0);

        streamCallback.streamAdvertised(streamId, name, streamType, address, port);
    }
}
